src/core/DomainDecomposition.cpp

```diff
@@ -270,19 +272,29 @@ void DomainDecomposition::share() {
       if( actions[i]->unique_local_needs_update ) actions[i]->updateUniqueLocal( !(dd && shuffledAtoms>0), g2l );
     }
     // Now reset unique for the new step
-    gch::small_vector<const std::vector<AtomNumber>*,32> vectors;
-    vectors.reserve(actions.size());
+    gch::small_vector<const std::vector<AtomNumber>*,32> forced_vectors;
+    gch::small_vector<const std::vector<AtomNumber>*,32> nonforced_vectors;
+    forced_vectors.reserve(actions.size()); nonforced_vectors.reserve(actions.size());
     for(unsigned i=0; i<actions.size(); i++) {
       if(actions[i]->isActive()) {
         if(!actions[i]->getUnique().empty()) {
           // unique are the local atoms
-          vectors.push_back(&actions[i]->getUniqueLocal());
+          if( actions[i]->actionHasForces() ) forced_vectors.push_back(&actions[i]->getUniqueLocal());
+          else nonforced_vectors.push_back(&actions[i]->getUniqueLocal());
         }
       }
     }
-    if(!vectors.empty()) atomsNeeded=true;
-    unique.clear();
-    mergeVectorTools::mergeSortedVectors(vectors.data(),vectors.size(),unique);
+    if( !(forced_vectors.empty() && nonforced_vectors.empty()) ) atomsNeeded=true;
+    // Merge the atoms from the atoms that have a force on
+    forced_unique.clear();
+    mergeVectorTools::mergeSortedVectors(forced_vectors.data(),forced_vectors.size(),forced_unique);
+    // Merge the atoms from the actions that don't apply forces
+    std::vector<AtomNumber> nonforced_unique;
```

Comment:
One thing that might be suboptimal here is that you do 3 merges:
- build forced_unique
- build nonforced_unique (local variable, then discarded)
- merge them in unique (merge of two vectors)

Maybe the following is faster:
- build forced_unique
- merge all the other actions with force_unique to build unique, in a single step

Notice that the merge tool receive a vector (or small_vector) of pointers, so you can easily make it point to the list you already made (forced_unique).

This would result in one less merge. Notice that the single merge that you are avoiding is the one with two lists only, which is more expensive than the other ones (these are the largest lists, and the priority queue can do little with two elements). But I am just speculating.

Reply:
Something like:
```
nonforced_vectors.push_back(&force_unique);
mergeVectorTools::mergeSortedVectors(nonforced_vectors.data(),nonforced_vectors.size(),unique);
```
and you are done

Reply:
Or it could be a good idea to use this for the last (pairwise) merge, in your current implementation: https://en.cppreference.com/w/cpp/algorithm/set_union . It could be faster than the priority queue because it does a direct comparison of two elements at a time, without the book-keeping